docs(guides): add debugging guide (#1174)
ionic-team/stencil-site: docs/guides/vs-code-debugging.md

---
title: Debugging With VS Code
sidebar_label: VS Code Debugging
description: How to debug a Stencil component using VS Code
slug: /vs-code-debugging
---

# Debugging With VS Code

VS Code offers a streamlined debugging experience that can be started with a single click when using [launch configurations](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations).

> For more information on how to use the VS Code debugger, please visit the [debugger documentation](https://code.visualstudio.com/docs/editor/debugging).

Reviewer: This sentence can read a little funky - since folks are coming here to figure out how to use the debugger with Stencil, we might want to change this to specify the link is for general usage of a debugger:
```suggestion
If you're unfamiliar with using the VS Code debugger in general, please visit the [VS Code debugger documentation](https://code.visualstudio.com/docs/editor/debugging) for a primer.
```

Author: Yup, like how that sounds

## Requirements for Debugging

> In order for a debugger to function, the Stencil project must be configured to generate source maps for the compiled web components back to the source code. As of Stencil v3, source maps are generated by default, but be sure to double check the project's Stencil config does not disable this behavior. More information regarding source maps in Stencil can be found in the [Stencil documentation](https://stenciljs.com/docs/config#sourcemap).

Reviewer: Minor, can we use absolute paths here?

```suggestion
In order for a debugger to function, the Stencil project must be configured to generate source maps for the compiled web components back to the source code. As of Stencil v3, source maps are generated by default, but be sure to double check the project's Stencil config does not disable this behavior. More information regarding source maps in Stencil can be found in the [project configuration documentation](../config/01-overview.md#sourcemap).
```

> In order for a debugger to function, the Stencil project must be configured to generate source maps for the compiled web components back to the source code. As of Stencil v3, source maps are generated by default, but be sure to double check the project's Stencil config does not disable this behavior. More information regarding source maps in Stencil can be found in the [Stencil documentation](https://stenciljs.com/docs/config#sourcemap).

Reviewer: Spelling nit:
```suggestion
In order for a debugger to function, the Stencil project must be configured to generate source maps for the compiled web components back to the source code. As of Stencil v3, source maps are generated by default, but be sure to double-check the project's Stencil config does not disable this behavior. More information regarding source maps in Stencil can be found in the [Stencil documentation](https://stenciljs.com/docs/config#sourcemap).
```

## Debugging Stencil Components In a Web App

It's a common use case to want to step through a web component's code as it executes in the browser and VS Code makes that process simple. Combining the debugger with Stencil's dev server in watch mode will allow you to debug changes as they're made.

### Configuring the VS Code Debugger

> To debug Stencil components as they run in a browser (web app), create (or edit) the `launch.json` file with the following configuration:

Reviewer: Should we clarify this file is often found in `.vscode/` in a project's root?
```suggestion
To debug Stencil components as they run in a browser (web app), create (or edit) the `.vscode/launch.json` file with the following configuration:
```

Author: Good call

> ```json title="launch.json"

Reviewer: Similar to my point above, should we clarify this file is often found in `.vscode/` in a project's root?
```suggestion
```json title=".vscode/launch.json"
```

{
  ...,
  "configurations": [
    ...,
    {
      "type": "chrome",
      "request": "launch",
      "name": "Launch Chrome against localhost",
      "url": "http://localhost:3333",
      "sourceMaps": true,
      "sourceMapPathOverrides": {
        "*": "${webRoot}/*"
      }
    }
  ]
}
```

:::note
If your Stencil project is within a monorepo structure, you may want (or need) to add the `webRoot` option to the above config to point to the Stencil project's directory. For instance, if you have your Stencil project at `/packages/stencil-library`, you would add the following to the config:

```json
{
    ...,
    "webRoot": "${workspaceFolder}/packages/stencil-library",
}
```

> :::

Reviewer: ![Screenshot 2023-07-14 at 1 50 43 PM](https://github.com/ionic-team/stencil-site/assets/1930213/eb022367-ba04-4de7-81f5-34c31ff01f86)

I _think_ adding the JSON block here to the node is intentional. IDK if we need the newline that precedes it, but maybe that helps clarify that this block ender is a part of the `note` above?

> :::

Author: Yes, JSON was intentional. The extra lines were just how prettier auto-formatted it 🤷‍♂️
